### PR TITLE
Add a way to get unmapped reads from a BAM.

### LINKED ISSUE #133: Add a way to get unmapped reads from a BAM
It's possible to filter by interval, but there's currently no way to include unmapped reads in the filter.

Needed for https://github.com/broadinstitute/gatk/issues/2571.

### PR DESCRIPTION
**Pull request by tomwhite**
Fixes #133.

### CODE REVIEW

File: src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java [hunk 1]
```diff
@@ -72,31 +74,61 @@
 	// set this to true for debug output
 	public final static boolean DEBUG_BAM_SPLITTER = false;
 
+	static final String BOUNDED_TRAVERSAL_PROPERTY = "hadoopbam.bam.bounded-traversal";
```
Comment (heuermh) [inline src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java]: Should this be `public static` as `INTERVALS_PROPERTY` is?  Also needs javadoc.

File: src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java [hunk 2]
```diff
@@ -72,31 +74,61 @@
...
+
 	/**
 	 * Filter by region, like <code>-L</code> in SAMtools. Takes a comma-separated
 	 * list of intervals, e.g. <code>chr1:1-20000,chr2:12000-20000</code>. For
 	 * programmatic use {@link #setIntervals(Configuration, List)} should be preferred.
 	 */
 	public static final String INTERVALS_PROPERTY = "hadoopbam.bam.intervals";
 
+	static final String TRAVERSE_UNPLACED_UNMAPPED_PROPERTY = "hadoopbam.bam.traverse-unplaced-unmapped";
```
Comment (heuermh) [inline src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java]: Should this be `public static` as `INTERVALS_PROPERTY` is?  Also needs javadoc.

File: src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java [hunk 3]
```diff
@@ -72,31 +74,61 @@
...
-		StringBuilder sb = new StringBuilder();
-		for (Iterator<T> it = intervals.iterator(); it.hasNext(); ) {
-			Locatable l = it.next();
-			sb.append(String.format("%s:%d-%d", l.getContig(), l.getStart(), l.getEnd()));
-			if (it.hasNext()) {
-				sb.append(",");
+		setTraversalParameters(conf, intervals, false);
+	}
+
+	public static <T extends Locatable> void setTraversalParameters(Configuration conf,
```
Comment (heuermh) [inline src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java]: Needs javadoc.

File: src/test/java/org/seqdoop/hadoop_bam/TestBAMInputFormat.java [hunk 4]
```diff
@@ -26,11 +26,12 @@
   private TaskAttemptContext taskAttemptContext;
   private JobContext jobContext;
   
-  private void completeSetup(List<Interval> intervals) {
+  private void completeSetup(boolean boundedTraversal, List<Interval> intervals, boolean
```
Comment (heuermh) [inline src/test/java/org/seqdoop/hadoop_bam/TestBAMInputFormat.java]: I know this is just test code, but with two boolean parameters it is not really clear in the test methods below which situation is under test.

How about adding new methods `setupBoundedTraversalUnplacedUnmapped(List<Interval>)` and similar that delegate to this one?

**Comment by tomwhite** [comment]
@heuermh thanks for the review. I have addressed all your points: I added documentation to all methods, and also entries in the README to cover bounded traversal. I made the test code clearer, and also added a negative test for the case of mapped only reads (which are not currently supported).

Please let me know if it looks OK now.

**Pull request merged**